### ISSUE TITLE
Switch Component

### ISSUE THREAD

**Issue by gliechtenstein** [Jasonette/Jasonette#1]
Going forward, we'll be using this repo as the place to discuss JSON specs for the features we have decided to build.

Let's start this out with "switch", a component with boolean value. More details on this by @maks over here: https://github.com/Jasonette/documentation/issues/79

Agenda:

1. Need to research both iOS and Android to see what features are available for each platform.
2. Then need to come up with a JSON syntax to express those

**Comment by maks** [comment]
Thanks @gliechtenstein 
Moving the info I had in the documentation repo to here:

On iOS this would be a [switch](https://developer.apple.com/ios/human-interface-guidelines/ui-controls/switches/)

On modern Android this is also now a [switch](https://developer.android.com/guide/topics/ui/controls/togglebutton.html) (historically in older, pre v4 versions they were "togglebuttons").

**Comment by ghost** [comment]
I think you could simply look at the [Reach switch](https://facebook.github.io/react-native/docs/switch.html), that's very widely used and simple. Attaching an action to a component is essentially registering a callback, so React's Switch's onValueChange callback becomes the jason "action", I would define two, one for each state rather than one that's parametererized because there's no way in jason to specify a handler with parameters. Similar to error/success, this would be on/off

**Comment by gliechtenstein** [comment]
I think there are two variations for boolean components:

| switch | checkbox
| -------|--------------
![ios7_switches](https://user-images.githubusercontent.com/16613330/29550927-68cf1d5c-86de-11e7-8c82-a593fc373d9a.png) | ![check](https://user-images.githubusercontent.com/16613330/29550991-ada2bf2e-86de-11e7-9108-862c01cc80a6.jpg)


---


Personally I prefer checkbox because it's 100% clear if it's selected or not, whereas using a switch in most cases is not at all clear and I think it's a bad UX. The ambiguity is the reason why we need an additional caption to indicate if it's on or off (as proposed over at https://github.com/Jasonette/documentation/issues/79 as `on_text` and `off_text` attributes)

This would make the syntax much simpler since the component is nothing more than a single element with on/off state, and no need for separate color/text for each of on/off states.

But this is just me, would love to hear what everyone else thinks.

**Comment by clsource** [comment]
@darenr This component is also interesting to look at https://github.com/bradmartin/nativescript-checkbox 👍 . It can be used as a multiple selection checkbox or radio buttons (single selection).

**Comment by maks** [comment]
@gliechtenstein I had in mind the switch not checkbox though I do agree that the checkbox UX is more clear. Unfort it's the checkbox that is used on both iOS and Android for "System Settings" UI and seems widely used else where so I'd prefer it more on the basis of consistency when implementing app preferences.
But I'd be happy if we could implement both switch **and** checkbox as they both have thier uses.

**Comment by maks** [comment]
@darenr I agree with you about the lack of a way to specify handler with parameters, but I think that should be solved in general rather than just specifically for the switch component. I'll raise that in a separate issue here.

**Comment by clsource** [comment]
@maks 
So 
switch is for on/off needs. Returns a single value
And checkboxes are for multiple selection or single selection (radio buttons). Returns an object with key => value.

**Comment by ghost** [comment]
I have to admit after years with iOS, I _still_ get confused with the boolean switch (I have the "1" enabled to make it clearer).

Since checkbox includes the bool case as a subset but provides more, I'd vote for checkboxes.

**Comment by clsource** [comment]
Some ideas. It's working like a textfield

Switch 
returns a boolean value.
On = True (1)
Off = False (0)

```
{
"type" : "switch",
"name" : "unique_name",
"style" " {
// switch styles
},
"options" : {
 "default" : "label when off",
 "text" : "label when on",
 "checked" : "true|false"
}
  
```

could be accesed as `{{$get.unique_name}}`

Checkbox
Returns a boolean value or a custom value when checked.
```
{
"type" : "checkbox",
"name" : "unique_name", // this will be shared by all checks
"options" : {
  "class" : "unique_in_namespace",
  "default" : "optional value if not checked (false)",
  "value" : "optional value if checked (true)",
 "checked" : "true|false"
}
```
Usage.
`{{$get.unique_name}}` will return all the checkbox values as an object
`{{$get.unique_name.unique_in_namespace}}` will return a single checkbox object.

Radio
Similar to Checkbox. But only one value can be selected at once.
Returns a boolean value or a custom value when checked.
```
{
"type" : "radio",
"name" : "namespace", // this will be shared by all checks
"options" : {
  "class" : "unique_in_namespace",
  "default" : "optional value if not checked (false)",
  "value" : "optional value if checked (true)",
 "checked" : "true|false"
}
```
Usage.
`{{$get.unique_name}}` will return all the checkbox values as an object
`{{$get.unique_name.unique_in_namespace}}` will return a single checkbox object.
`{{$get.unique_name.selected}}` will return the selected checkbox object

**Comment by maks** [comment]
@clsource I like that but I think we still want an action too to be able to trigger something when the value changes

**Comment by clsource** [comment]
@maks Yes I think an action could be put in switch, checkbox and radios when a valueChangedEvent is detected. So you can call an API Rest and send the values obtained in the `{{$get.name}}` var 👍

**Comment by clsource** [comment]
I think this component should be separated in 3 different components. For each use case. So they have it's own spec discussion.

- Switch Component (True/False) Only
- Checkbox Component (Multi value)
- Radio Component (Single Selection Multi Value)

**Comment by gliechtenstein** [comment]
We need to prioritize the resources so let's just pick one and go with it for now. We can decide to add more later if we find that they're really necessary.

Anyway, as much as I personally hate `switch` component for the reasons I mentioned above, I think we should go with this since among all the options that have been brought up, `switch` is the only component that's native on both iOS and Android.

Also I believe in keeping the components as minimal as possible so let's not add many attributes if we don't need to.

Thankfully Google seems to have come to the same realization (See [Material design guide for switch](https://material.io/guidelines/components/selection-controls.html#selection-controls-switch))

What I think are needed are:

- **name:** will be used to set the local variable
- **value:** preset value (`"1"` if on, `"0"` if off). If not specified, it's `"0"` by default.
- **style**
    - **"color":** active state color
    - **"color:disabled":** inactive state color (use default color if not specified)
- **action:** [action](http://docs.jasonette.com/actions/) to trigger

It's similar to how [slider](http://docs.jasonette.com/components/#slider) works and I think that's good enough for now. Let me know what you think.

**Comment by clsource** [comment]
Great thing that Google deprecated the messages for on/off.

So the spec would be something like this I think

```
"type" : "switch",
"name" : "local-variable-name",
"value" : "true|false|1|0" // this is a boolean, default is false (0),
"style" : {
   "color" : "active state color" (default os active state color if not specified),
   "color:disabled" : "inactive state color (default os inactive state color if not specified)
},
"action" : action trigger
```
